packages/fetchai/connections/gym/connection.py

```diff
@@ -58,6 +61,55 @@ def __init__(self, address: Address, gym_env: gym.Env):
             self.THREAD_POOL_SIZE
         )
         self.logger: Union[logging.Logger, logging.LoggerAdapter] = logger
+        self._dialogues = GymDialogues(self.address)
+
+    def _get_message_and_dialogue(
+        self, envelope: Envelope
+    ) -> Tuple[GymMessage, Optional[GymDialogue]]:
+        """
+        Get a message copy and dialogue related to this message.
+
+        :param envelope: incoming envelope
+
+        :return: Tuple[MEssage, Optional[Dialogue]]
+        """
+        message = cast(GymMessage, envelope.message)
+        message = copy.deepcopy(
+            message
+        )  # TODO: fix; need to copy atm to avoid overwriting "is_incoming"
+        message.is_incoming = True  # TODO: fix; should be done by framework
+        message.counterparty = envelope.sender  # TODO: fix; should be done by framework
+        dialogue = cast(GymDialogue, self._dialogues.update(message))
+        if dialogue is None:  # pragma: nocover
+            logger.warning("Could not create dialogue for message={}".format(message))
+        return message, dialogue
+
+    @staticmethod
+    def _set_response_dialogue_references(
+        response_message: GymMessage,
+        incoming_message: GymMessage,
+        dialogue: GymDialogue,
+    ) -> GymMessage:
+        response_message.set(
+            "dialogue_reference", dialogue.dialogue_label.dialogue_reference
+        )
+        """
+        Create a response message  with all dialog ue details.
+
+        :param response_message: message to be sent
+        :param incoming_message: message that response constructed for
+        :param dialogue: a dialog for messages
+
+        :return: new response message with all dialogue details.
+        """
+        response_message.set("target", incoming_message.message_id)
+        response_message.set("message_id", incoming_message.message_id + 1)
+        response_message.counterparty = incoming_message.counterparty
+        dialogue.update(response_message)
+        response_message = copy.deepcopy(
+            response_message
+        )  # TODO: fix; need to copy atm to avoid overwriting "is_incoming"
+        return response_message
```

Comment:
`target, message_id and dialogue_reference` should be set on message construction (line 114 onwards), not here

Reply:
done